site/source/docs/porting/exceptions.rst

```diff
@@ -79,14 +83,41 @@ be supported in all WebAssembly engines yet
 Debugging Exceptions
 ====================
 
-When :ref:`ASSERTIONS <debugging-ASSERTIONS>` is enabled, uncaught exceptions
-will print stack traces for debugging. :ref:`ASSERTIONS <debugging-ASSERTIONS>`
-is enabled by default in :ref:`-O0 <emcc-O0>` and disabled in optimized builds
-(:ref:`-O1 <emcc-O1>` and above). You can enable it by passing ``-sASSERTIONS``
-to the ``emcc`` command line in optimized builds as well.
-
-You can also catch and examine C++ exceptions from JavaScript. See
-:ref:`handling-c-exceptions-from-javascript`.
+Stack Traces
+------------
+
+For :ref:`native Wasm exceptions
+<webassembly-exception-handling-based-support>`, when :ref:`ASSERTIONS
+<debugging-ASSERTIONS>` is enabled, uncaught exceptions will print stack traces
+for debugging. :ref:`ASSERTIONS <debugging-ASSERTIONS>` is enabled by default
+in :ref:`-O0 <emcc-O0>` and disabled in optimized builds (:ref:`-O1 <emcc-O1>`
+and above). You can enable it by passing ``-sASSERTIONS`` to the ``emcc``
+command line in optimized builds as well. To display Wasm function names in
+stack traces, you also need :ref:`-g <emcc-g>` or :ref:`-gsource-map
+<emcc-gsource-map>`.
+
+In JavaScript, you can also examine the stack traces using
+`WebAssembly.Exception.prototype.stack
+<https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Exception/stack>`_
+property. For example:
+
+.. code-block:: javascript
+
+  try {
+    ... // some code that calls WebAssembly
+  } catch (e) {
+    // Do something with e.stack
+    console.log(e.stack);
+  }
+
+Stack traces within Wasm code are not supported in :ref:`JavaScipt-based
```

Comment:
I'm not sure what you mean here.  Surly you can catch emscripten EH exceptions and see their stack trace can't you?

Reply:
I don't think so?

```console
aheejin@aheejin:~/test/stack$ emcc -g -fexceptions test.cpp -o test.js -sENVIRONMENT=shell,node

aheejin@aheejin:~/test/stack$ node --trace-uncaught test.js

/usr/local/google/home/aheejin/test/stack/test.js:144
      throw ex;
      ^
66800
Thrown at:
    at /usr/local/google/home/aheejin/test/stack/test.js:144:7
    at emit (node:events:526:28)
    at node:internal/process/execution:167:25

Node.js v17.5.0

aheejin@aheejin:~/test/stack$ d8 test.js
exiting due to exception: 66800
```
Node only shows the JS part and d8 doesn't show anything.

JS exception library uses `makeThrow`, which just throws a number: https://github.com/emscripten-core/emscripten/blob/90cccdf2798c22b4fa3f82f586a57147750019de/src/parseTools.js#L614

Unless you create an `Error` object, it doesn't have stack traces. But if we make it an `Error`, all JS exception related library functions, based on that the thrown thing is a number (an address, to be precise), wouldn't work.

Reply:
FWIW, draft PR #18535 would fix that.

Reply:
Oh, I didn't realize that throwing a number doesn't give you a stack trace.